text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,755 @@
+        if self.is_value::<T>() {
+            // safety: If `self.is_value::<T>()` returns true, `response` must
+            // be of the correct type. This is enforced by the private `type_id`
+            // field.
+            Some(unsafe { &mut *(&mut self.response as *mut _ as *mut ValueResponse<T>) })
+        } else {
+            None
+        }
+    }
+
+    /// Provides a reference of type `&'a T` in response to this request.
+    ///
+    /// If a reference of type `&'a T` has already been provided for this
+    /// request, or if the request is for a different type, this call will be
+    /// ignored.
+    ///
+    /// This method can be chained within `provide` implementations to concisely
+    /// provide multiple objects.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use object_provider::{ObjectProvider, Request};
+    /// # use std::fmt;
+    /// struct MyProvider {
+    ///     name: String,
+    /// }
+    ///
+    /// impl ObjectProvider for MyProvider {
+    ///     fn provide<'a>(&'a self, request: &mut Request<'a>) {
+    ///         request
+    ///             .provide_ref::<Self>(&self)
+    ///             .provide_ref::<String>(&self.name)
+    ///             .provide_ref::<str>(&self.name)
+    ///             .provide_ref::<dyn fmt::Display>(&self.name);
+    ///     }
+    /// }
+    /// ```
+    pub fn provide_ref<T: ?Sized + 'static>(&mut self, value: &'a T) -> &mut Self {
+        self.provide_ref_with(|| value)
+    }
+
+    /// Lazily provides a reference of type `&'a T` in response to this request.
+    ///
+    /// If a reference of type `&'a T` has already been provided for this
+    /// request, or if the request is for a different type, this call will be
+    /// ignored.
+    ///
+    /// The passed closure is only called if the value will be successfully
+    /// provided.
+    ///
+    /// This method can be chained within `provide` implementations to concisely
+    /// provide multiple objects.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use object_provider::{ObjectProvider, Request};
+    /// # fn expensive_condition() -> bool { true }
+    /// struct MyProvider {
+    ///     a: String,
+    ///     b: String,
+    /// }
+    ///
+    /// impl ObjectProvider for MyProvider {
+    ///     fn provide<'a>(&'a self, request: &mut Request<'a>) {
+    ///         request.provide_ref_with::<String, _>(|| {
+    ///             if expensive_condition() {
+    ///                 &self.a
+    ///             } else {
+    ///                 &self.b
+    ///             }
+    ///         });
+    ///     }
+    /// }
+    /// ```
+    pub fn provide_ref_with<T: ?Sized + 'static, F>(&mut self, cb: F) -> &mut Self
+    where
+        F: FnOnce() -> &'a T,
+    {
+        if let Some(RefResponse(response @ None)) = self.downcast_ref_response::<T>() {
+            *response = Some(cb());
+        }
```

Comment:
Surely there's a typo here somewhere? `response @ None` can't be doing anything interesting, and it doesn't mean anything to try to deref it in the next line. Maybe the second line is meant to be `*self.response =`, but then I have no clue what the purpose of the `@` pattern is.

Reply:
Does this help explain it?

```rust
struct Data(Option<()>);

let mut data = Data(None);
if let Some(Data(empty @ None)) = Some(&mut data) {
    *empty = Some(());
}
assert_eq!(data.0, Some(()));
```

Reply:
Oof, the default binding modes threw me for a loop here. I thought they only worked at the top level of each pattern, it didn't occur to me that `response @ None` could be desugaring to `response @ &mut None`.